### PR TITLE
Support ArgMin and ArgMax for NVGPU Backend 

### PR DESCRIPTION
**Pull request by AyanmoI**
Implement argmax and argmin ops using cudnn reduceTensor APIs.  Add additional unit tests for argmax and argmin.

### CODE REVIEW

File: src/ngraph/runtime/gpu/unit_test.manifest [hunk 1]
```diff
@@ -34,6 +34,9 @@ max_pool_3d
 avg_pool_3d
 argmin_trivial
 argmax_trivial
+argmax_3D_axis_0
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/unit_test.manifest]: You are disabling these new tests. Do they not pass? Do they pass for all other backends?

**Comment by AyanmoI** [comment]
I added these tests mainly to get a better understanding of how the Argmax op works in higher order tensors.
I am just temporarily disabling these tests for the GPU since I have not completed my implementation. I will enable them once I have a working implementation.

File: src/ngraph/runtime/gpu/cublas_emitter.cpp [hunk 2]
```diff
@@ -211,6 +211,16 @@ size_t runtime::gpu::CUBLASEmitter::build_dot(const element::Type& dtype,
     return primitive_index;
 }
 
+size_t runtime::gpu::CUBLASEmitter::build_argmax(const element::Type& dtype,
+                                                 const Shape& arg0_shape,
+                                                 const Shape& out_shape,
+                                                 size_t axis,
+                                                 const Node* node)
+{
+    size_t primitive_index = 0;
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/cublas_emitter.cpp]: I think 0 could already been used. you can not just return 0.
 You can generate one with same process as others and use it with an empty implementation.

File: test/backend_test.in.cpp [hunk 3]
```diff
@@ -9371,6 +9371,66 @@ NGRAPH_TEST(${BACKEND_NAME}, argmax_trivial)
...
+    Shape rshape{4, 2};
+    auto A = make_shared<op::Parameter>(element::f32, shape);
+    auto f =
+        make_shared<Function>(make_shared<op::ArgMax>(A, 0, element::i32), op::ParameterVector{A});
+
+    auto backend = runtime::Backend::create("${BACKEND_NAME}");
+
+    // Create some tensors for input/output
+    auto a = backend->create_tensor(element::f32, shape);
+    copy_data(a, vector<float>{8, 4, 12, 10, 2, 9, 1, 5, 6, 7, 11, 3,
```
Comment (fengleitian) [inline test/backend_test.in.cpp]: if you want, you can lay out 342 like this
```
xx
xx
xx
xx

xx
xx
xx
xx

xx
xx
xx
xx
```
here is an example:
https://github.com/NervanaSystems/ngraph/blob/2b289df09692b0989873229ed745787ea8881647/test/backend_test.in.cpp#L5159

easier to read.

Comment (fengleitian) [reply test/backend_test.in.cpp]: same of other tests.

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp [hunk 4]
```diff
@@ -170,35 +172,79 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
...
-        new gpu::primitive{[=, &desc, &input_desc, &output_desc](void** inputs, void** outputs) {
+    std::unique_ptr<gpu::primitive> reduce;
+    switch (reduction_mode)
+    {
+    case ReductionMode::Reduce:
+    {
+        // emit reduce operation
+        reduce.reset(new gpu::primitive{
+            [=, &desc, &input_desc, &output_desc](void** inputs, void** outputs) {
+                CUDNN_SAFE_CALL(cudnnSetReduceTensorDescriptor(desc,
```
[the remark this answers is not in the document]
Comment (AyanmoI) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: makes complete sense...Done!

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp [hunk 5]
```diff
@@ -170,35 +172,79 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
...
+                                                  beta,
+                                                  output_desc,
+                                                  outputs[0]));
+                debug_sync();
+            }});
+
+        break;
+    }
+
+    case ReductionMode::ArgMax_ArgMin:
```
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: Is it possible that we just have one primitive that always produces indices, and for argmin/max the outputs[0] is in a different place of the cudnnReduceTensor call?

```
void* reduce_buffer = (reduction_mode == ReductionMode::Reduce) ? outputs[0] : reduce_buffer;
```

File: test/backend_test.in.cpp [hunk 6]
```diff
@@ -9371,6 +9419,168 @@ NGRAPH_TEST(${BACKEND_NAME}, argmax_trivial)
...
+
+                                       {{8, 4}, //ch2
+                                        {6, 1},
+                                        {5, 3},
+                                        {11, 7}}})
+                  .get_vector());
+    auto result = backend->create_tensor(element::i32, rshape);
+
+    backend->call_with_validate(f, {result}, {a});
+    EXPECT_EQ((test::NDArray<int, 2>({{0, 0, 1, 1},  //
```
Comment (csullivan) [inline test/backend_test.in.cpp]: remove `//`

File: test/backend_test.in.cpp [hunk 7]
```diff
@@ -9371,6 +9419,168 @@ NGRAPH_TEST(${BACKEND_NAME}, argmax_trivial)
     EXPECT_EQ((vector<int>{1, 3, 0}), read_vector<int>(result));
 }
 
+NGRAPH_TEST(${BACKEND_NAME}, argmax_3D_axis_0) // Along Channels
```
Comment (csullivan) [inline test/backend_test.in.cpp]: Nice tests, thanks for adding more coverage.

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: IMO it comes down to whether there is a non-trivial performance difference? We don't want to always produce them if there is.

File: test/backend_test.in.cpp (hunk 6 above)
Comment (AyanmoI) [reply test/backend_test.in.cpp]: Had to add these to stop clang from putting all the rows in one line. This is just for visualization purpose. I can remove them.

Comment (csullivan) [reply test/backend_test.in.cpp]: Ah got it, it's fine by me. Maybe, in that case, make a comment about each block as in the other tests?

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp [hunk 8]
```diff
@@ -170,35 +172,79 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
     CUDNN_SAFE_CALL(cudnnGetReductionWorkspaceSize(
         *m_ctx->cudnn_handle, desc, input_desc, output_desc, &workspace_size));
     size_t workspace_idx = allocator.reserve_workspace(workspace_size);
```
[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: is it easy to add in `gpu_memory_manager`?

[the remark this answers is not in the document]
Comment (AyanmoI) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: Makes sense but I'm not sure what testcase I might use to test this

[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: I suppose we could also register a memory primitive that returns nullptr when size=0 and avoid changes to user code. @AyanmoI I suspect the returned workspace size is 0 for a small reduction, e.g. the vector to scalar.

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp [hunk 9]
```diff
@@ -170,35 +172,79 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
...
+        size_t reduce_buffer_idx =
+            allocator.reserve_workspace(shape_size(output_shape) * reduction_output_type_size);
+        reduce.reset(new gpu::primitive{[=, &desc, &input_desc, &output_desc](void** inputs,
+                                                                              void** outputs) {
             CUDNN_SAFE_CALL(cudnnSetReduceTensorDescriptor(desc,
                                                            reduce_op,
                                                            data_type,
                                                            CUDNN_NOT_PROPAGATE_NAN,
-                                                           CUDNN_REDUCE_TENSOR_NO_INDICES,
+                                                           CUDNN_REDUCE_TENSOR_FLATTENED_INDICES,
```
[the remark this answers is not in the document]
Comment (AyanmoI) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: To gt cudnn indices, we could use pooling backprop to obtain the indices

[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: Right, but they need to evaluated in forward pooling anyway, and if said information was available to the backprop, we could avoid the backprop evaluation. e.g. op::MaxpoolWithIndices in the IA transformer. Anyway, separate discussion from this PR, just wanted to ask about it.

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: I suspect there is only a small difference in performance, but the memory overhead is not worth it. Let's keep it as you have it.

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (AyanmoI) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: I will address this in a separate PR

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp [hunk 10]
```diff
@@ -170,35 +177,78 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
...
+                                                  outputs[0]));
+                debug_sync();
+            }});
+
+        break;
+    }
+
+    case ReductionMode::ArgReduce:
+    {
+        size_t indices_size = shape_size(output_shape) * sizeof(int);
```
Comment (gcwenger) [inline src/ngraph/runtime/gpu/cudnn_emitter.cpp]: sizeof(int) -> sizeof(uint32_t) (to more precisely correspond to CUDNN_32BIT_INDICES)

[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: nice catch.

[the remark this answers is not in the document]
Comment (gcwenger) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: The trick is that use of cudnnReduceTensor only works for uint32_t.

Comment (AyanmoI) [reply src/ngraph/runtime/gpu/cudnn_emitter.cpp]: Resolved by adding error check as discussed

File: src/ngraph/runtime/gpu/cudnn_emitter.cpp [hunk 11]
```diff
@@ -216,9 +218,15 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
...
-            allocator.reserve_workspace(shape_size(output_shape) * dtype.size());
+        if (output_type != element::i32)
+        {
+            std::stringstream ss_er;
+            ss_er << "Unsupported Type: Only uint32 supported for indices in op ArgReduce ";
+            throw std::invalid_argument(ss_er.str());
+        }
+
+        size_t indices_size = shape_size(output_shape) * output_type.size();
+        size_t reduce_buffer_idx = allocator.reserve_workspace(input_type.size());
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/cudnn_emitter.cpp]: `allocator.reserve_workspace(~~input_type.size()~)` need `shape_size(output_shape) * input_type.size()`

**Pull request merged**